overlord/snapstate/snapstate.go

```diff
@@ -532,6 +532,24 @@ func doInstall(st *state.State, snapst *SnapState, snapsup *SnapSetup, flags int
 		prev = installHook
 	}
 
+	if snapsup.QuotaGroupName != "" {
+		if err := CheckQuotaChangeConflictMany(st, []string{snapsup.QuotaGroupName}); err != nil {
+			return nil, err
+		}
+
+		// This could result in doing 'setup-profiles' twice, but
+		// unfortunately we can't execute this code earlier as the snap
+		// needs to appear as installed first.
+		// XXX: Could we maybe move 'setup-profiles' to after this somehow? Or
+		// maybe just accept this could incur 'setup-profiles' twice?
+		quotaControlTask := st.NewTask("quota-add-snap", fmt.Sprintf(i18n.G("Add snap %q%s to quota group %q"),
+			snapsup.InstanceName(), revisionStr, snapsup.QuotaGroupName))
+		quotaControlTask.Set("snap-name", snapsup.InstanceName())
+		quotaControlTask.Set("quota-name", snapsup.QuotaGroupName)
```

Comment:
it feels that this code belong into a QuotaAddSnapTask helper set from servicestate into snapstate, has it happens with some of the hook stuff, see SetupInstallHook  etc for example

at that point the conflict stuff could move back there as well

Reply:
I did this as you mentioned, but I kept the conflict code seperated (but moved back to servicestate) and in similar style as the original conflict code, as I believe that makes more sense when we now have multiple places that need to check for conflicts.